api/rest/cluster_bundle_handler.go

```diff
@@ -0,0 +1,360 @@
+			write(w, jsonMarshal(bundle))
+			return
+		}
+	}
+
+	// we would only get here if we didn't find the bundle on any of the masters
+	writeJSONError(w, http.StatusNotFound, fmt.Errorf("bundle %s did not exist on any masters", id))
+}
+
+// Delete will delete a given bundle, proxying the call if the given bundle exists
+// on a different master
+func (c *ClusterBundleHandler) Delete(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+
+	masters, err := c.getMasterNodes()
+	if err != nil {
+		writeJSONError(w, http.StatusInternalServerError, fmt.Errorf("unable to get list of masters: %s", err))
+		return
+	}
+
+	ctx := context.Background()
+
+	found := false
+	for _, n := range masters {
+		err = c.client.Delete(ctx, n.baseURL, id)
+		if err != nil {
+			// some errors tell us the bundle was found on a master but something else was wrong so we end and return an error status
+			// but a NotFound error means it should keep going
+			switch err.(type) {
+			case *DiagnosticsBundleNotCompletedError:
+				w.WriteHeader(http.StatusNotModified)
+				return
+			case *DiagnosticsBundleUnreadableError:
+				writeJSONError(w, http.StatusInternalServerError, err)
+				return
+			}
+		}
+		if err == nil {
+			found = true
+		}
+	}
+
+	if !found {
+		writeJSONError(w, http.StatusNotFound, fmt.Errorf("bundle %s not found on any master", id))
+	}
+}
+
+// Download will download the given bundle, proxying the call to the appropriate master
+func (c *ClusterBundleHandler) Download(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+
+	// TODO: for this one specifically, it would be ideal to detect if the bundle exists on the calling master
+	// first since then we can skip the intermediate download
+	masters, err := c.getMasterNodes()
+	if err != nil {
+		writeJSONError(w, http.StatusInternalServerError, fmt.Errorf("unable to get list of masters: %s", err))
+		return
+	}
+
+	ctx := context.Background()
+
+	var masterWithBundle node
+	found := false
+	for _, n := range masters {
+		_, err = c.client.Status(ctx, n.baseURL, id)
+		if err != nil {
+			switch err.(type) {
+			case *DiagnosticsBundleUnreadableError:
+				writeJSONError(w, http.StatusInternalServerError, err)
+				return
+			case *DiagnosticsBundleNotCompletedError:
+				writeJSONError(w, http.StatusNotFound, err)
+				return
+			case *DiagnosticsBundleNotFoundError:
+				continue
+			}
+		} else {
+			masterWithBundle = n
+			found = true
+			break
+		}
+	}
+	if !found {
+		writeJSONError(w, http.StatusNotFound, &DiagnosticsBundleNotFoundError{id: id})
+		return
+	}
+
+	bundleDir, err := ioutil.TempDir("", "bundle-")
+	if err != nil {
+		writeJSONError(w, http.StatusInternalServerError, fmt.Errorf("error opening temp file to download bundle %s", err))
+		return
+	}
+	defer os.RemoveAll(bundleDir)
+
+	bundleFilename := filepath.Join(bundleDir, "bundle.zip")
+
+	err = c.client.GetFile(ctx, masterWithBundle.baseURL, id, bundleFilename)
```

Comment:
I think we need to redesign this part. The [doc string says the request is proxied](https://github.com/dcos/dcos-diagnostics/pull/131/files#diff-b272b368d5eab0a52ac82dff6f46a928R261) while the code downloads whole file and then serve it loading it all in memory. This could lead to serious problems with disk (when we are on node which owns the data we double the required space) and memory space (possible GBs of data).